### PR TITLE
Adds AOP logging style

### PR DESCRIPTION
**Pull request by nachomdo**
This PR is intended to solve https://github.com/FundingCircle/clj-loga/issues/6 :smile: 

- [x] Search target functions to hook
- [x] Hook target functions
- [x] Search target functions by default in the namespace where ```setup-loga``` is called
- [x] Make hooks play well with field filtering/obfuscation  
- [x] Update README explaining how to use hooks features

### CODE REVIEW

File: src/clj_loga/hooks.clj [hunk 1]
```diff
@@ -0,0 +1,39 @@
+(ns clj-loga.hooks
+  (:require [clj-loga.core :refer [log-wrapper]]
+            [clojure.set :refer [rename-keys]]
+            [robert.hooke :refer [add-hook]]))
```
Comment (antoniogarrote) [inline src/clj_loga/hooks.clj]: add-hook is not used in this ns, I think it has been moved to core.

Comment (nachomdo) [inline src/clj_loga/hooks.clj]: :+1:  :smile:  thanks!

File: src/clj_loga/core.clj [hunk 2]
```diff
@@ -102,12 +110,23 @@
...
+                (let [meta-args (-> (meta function) select-loga-keys (format-pre-log-msg args))]
+                  (prn "hook args" meta-args)
+                  (log-wrapper meta-args (apply f args)))))))
+
 (defn setup-loga
   "Initialize formatted logging."
-  [& {:keys [level obfuscate]
-      :or {level :info obfuscate []}}]
+  [& {:keys [level namespaces obfuscate]
+      :or {level :info namespaces [(default-namespace *ns*)] obfuscate []}}]
```
Comment (rsslldnphy) [inline src/clj_loga/core.clj]: Can we not just search all namespaces? Won't this prevent libraries from specifying `loga` tags?

File: src/clj_loga/hooks.clj [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+(defn target-functions-from-namespaces
+  "Gets a list of all the functions that contains loga's related metadata"
+  [namespaces]
+  (let [loga-meta-keys #{:clj-loga/operation :clj-loga/tag :clj-loga/pre-log-msg :clj-loga/post-log-msg}
+        all-functions (mapcat #(-> % ns-interns vals) namespaces)]
+    (filter #(some loga-meta-keys (keys (meta %))) all-functions)))
+
+(defn select-loga-keys [meta]
+  (-> meta (rename-keys {:clj-loga/tag :tag
```
Comment (rsslldnphy) [inline src/clj_loga/hooks.clj]: Might want to consider using properly namespaced keywords here (i.e. in the scope of the current namespace). Then users of the library can `require` the hooks namespace for less wordy metadata tags:

```clojure
(require [clj-loga.hooks :as h])

(defn my-handler
  {::h/tags [:foo :bar]}
  [args]
  (println args))
```

Only a nice to have though...

File: src/clj_loga/core.clj (hunk 2 above)
Comment (nachomdo) [inline src/clj_loga/core.clj]: I'm not sure. The idea was to let the user specify in which namespaces he wants to search so we can narrow down the final search for target functions. Also, I was thinking that sometimes we want to enable the logging for our project dependencies but other times we prefer them to remain in silence. We can default to search in all namespaces I was just afraid that maybe it's going to take too long to initialise the library. 

```Clojure
;; The user can choose from which libraries he wants to hear
(setup-loga :level :debug :namespaces ["project-ns.*" "shared-library-ns.*"]) 

;; As an alternative, he can default to search in all the namespaces
(setup-loga :level :debug)
```

File: src/clj_loga/hooks.clj (hunk 3 above)
Comment (nachomdo) [inline src/clj_loga/hooks.clj]: nice idea :+1: Thanks :smile:

File: src/clj_loga/core.clj (hunk 2 above)
Comment (rsslldnphy) [inline src/clj_loga/core.clj]: Fair enough. Maybe we could have an explicit `:all` option. I like your example though, with the provisor that we remove gender-specific pronouns from comments/docstrings ;-)

Comment (nachomdo) [inline src/clj_loga/core.clj]: :smile: :+1:

File: project.clj [hunk 4]
```diff
@@ -9,6 +9,7 @@
                  [com.taoensso/timbre "4.1.1"]                                   ;; Logging and profiling
                  [dire "0.5.3"]                                                  ;; Hooks
```
Comment (rsslldnphy) [inline project.clj]: @nachomdo I know this isn't one of your changes, but I don't think we're using `dire`, should we remove it?

Comment (nachomdo) [inline project.clj]: sure :smile:

**Comment by rsslldnphy** [comment]
Is anything preventing this getting merged? @apmaros? @timstott ? Are you happy with this PR?

File: src/clj_loga/core.clj [hunk 5]
```diff
@@ -95,19 +104,37 @@
...
+    m))
+
 (defn obfuscate-data [data keys-to-obfuscate]
   (postwalk
    (fn [element]
-     (if (map? element) (reduce obfuscate-key element keys-to-obfuscate)
-         element))
+     (cond
+       (map? element) (reduce obfuscate-key element keys-to-obfuscate)
+       (vector? element) (reduce obfuscate-vector-key element keys-to-obfuscate)
```
Comment (apmaros) [inline src/clj_loga/core.clj]: :+1: nice!

**Comment by apmaros** [comment]
@rsslldnphy All good with me. @nachomdo, can you just increase version to 4? Thanks

**Comment by timstott** [comment]
:+1: Awesome work! Can't wait to use it. @nachomdo @antoniogarrote please bump the version to 0.4.0 and we'll push it.

**Comment by nachomdo** [comment]
@timstott @apmaros  done! :smile:

**Pull request merged**